File: sdk/python/examples/with_payment.py
"""
Full x402 payment example.

Shows the complete flow:
1. Make request -> get 402
2. SDK automatically signs USDC payment on Base
3. Retries with X-PAYMENT header
4. Returns scraped content

Requirements:
  pip install "anybrowse[payment]"
  export ETH_PRIVATE_KEY="0x..."  # Must have USDC on Base

The wallet needs USDC (on Base network) to pay for API calls.
Typical cost is fractions of a cent per request.
"""

import os
import sys
from anybrowse import AnybrowseClient


def main():
    private_key = os.environ.get("ETH_PRIVATE_KEY")

    if not private_key:
        print("Set ETH_PRIVATE_KEY environment variable.")
        print("  export ETH_PRIVATE_KEY=0xYOUR_PRIVATE_KEY")
        sys.exit(1)

    client = AnybrowseClient(private_key=private_key)

    # Scrape a page with automatic payment
    print("Scraping https://example.com ...\n")
    try:
        result = client.scrape("https://example.com")

        print(f"Title:  {result.get(title)}")
        print(f"URL:    {result.get(url)}")
        print(f"Status: {result.get(status)}")
        print(f"\nMarkdown:\n{result.get(markdown, )[:500]}")

    except Exception as e:
        print(f"Error: {e}")

    # Search with payment
    print("\n\nSearching for web3 micropayments ...\n")
    try:
        serp = client.search("web3 micropayments", count=3)

        for r in serp.get("results", []):
            print(f"  - {r[title]}")
            print(f"    {r[url]}")
            print()

    except Exception as e:
        print(f"Error: {e}")

    # Crawl with payment
    print("Crawling x402 protocol specification ...\n")
    try:
        data = client.crawl("x402 protocol specification", count=2)

        for page in data.get("results", []):
            print(f"  [{page.get(status)}] {page.get(title)}")
            md = page.get("markdown", "")
            print(f"  {len(md)} chars of markdown")
            print()

    except Exception as e:
        print(f"Error: {e}")

    client.close()


if __name__ == "__main__":
    main()

File: sdk/typescript/examples/with-payment.ts
/**
 * x402 micropayment flow with the Anybrowse SDK.
 *
 * Paid endpoints (scrape, crawl, search) return HTTP 402 when payment is
 * required. This example shows how to catch that error and outlines the
 * payment signing flow.
 *
 * Run with: npx tsx examples/with-payment.ts
 */
import { AnybrowseClient, PaymentRequiredError } from "../src/index.js";

// x402 payment network constants
const PAYMENT = {
  chainId: 8453,                                           // Base
  asset: "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",    // USDC on Base
  payTo: "0x8D76E8FB38541d70dF74b14660c39b4c5d737088",    // Anybrowse recipient
} as const;

async function main() {
  // Initialize the client with payment configuration.
  // The privateKey is used to sign USDC transfer authorizations on Base.
  const client = new AnybrowseClient({
    payment: {
      privateKey: process.env.WALLET_PRIVATE_KEY ?? "",
    },
  });

  try {
    const result = await client.scrape("https://example.com");
    console.log("Scrape succeeded:", result.title);
  } catch (err) {
    if (err instanceof PaymentRequiredError) {
      console.log("Payment required by the server.");
      console.log("Payment details:", JSON.stringify(err.paymentDetails, null, 2));
      console.log();
      console.log("To complete the payment:");
      console.log(`  1. Approve USDC spend on Base (chain ${PAYMENT.chainId})`);
      console.log(`  2. Sign a transfer of the requested amount`);
      console.log(`     Asset:  ${PAYMENT.asset}`);
      console.log(`     Pay to: ${PAYMENT.payTo}`);
      console.log("  3. Include the signed payment header and retry the request");
      console.log();
      console.log(
        "Tip: Use an x402-compatible library (e.g. @coinbase/x402) to " +
        "automate payment signing and header injection."
      );
    } else {
      throw err;
    }
  }
}

main().catch(console.error);
